src/epic/distances/divergence_free.py

```diff
@@ -0,0 +1,217 @@
+from epic.modules import Residual
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+class DivergenceFree(pearson_mixin.PearsonMixin, base.Distance):
+    default_samples_cov = 500
+    default_samples_can = 500
+
+    def __init__(
+        self,
+        discount_factor: float,
+        state_sampler: Optional[samplers.BaseSampler[samplers.StateSample]] = None,
+        action_sampler: Optional[samplers.BaseSampler[npt.NDArray]] = None,
+        coverage_sampler: Optional[samplers.BaseSampler[samplers.CoverageSample]] = None,
+    ):
+        """Initialize the Divergence-Free Reward Distance.
+
+        Args:
+          state_sampler: The sampler for the state distribution. Optional if the coverage_sampler is provided.
+          action_sampler: The sampler for the action distribution. Optional if the coverage_sampler is provided.
+          coverage_sampler: The sampler for the coverage distribution. If not given,
+            a default sampler is constructed as drawing from the product
+            distribution induced by the distributions of state and action.
+          discount_factor: The discount factor.
+        """
+        if coverage_sampler is None:
+            assert (
+                state_sampler is not None and action_sampler is not None
+            ), "If no coverage sampler is given, state and action samplers must be provided."
+        else:
+            assert (
+                state_sampler is None and action_sampler is None
+            ), "If a coverage sampler is given, state and action samplers will not be used."
+
+        coverage_sampler = coverage_sampler or samplers.ProductDistrCoverageSampler(action_sampler, state_sampler)
+
+        super().__init__(discount_factor, state_sampler, action_sampler, coverage_sampler)
+
+        state_sample, _, _, _ = self.coverage_sampler.sample(1)
+        self.state_dim = int(np.prod(state_sample.shape))
+
+    def init_potential_net(self):
+        """Initializes the neural net that'll be used to trained to strip potentials from reward functions during canonicalization."""
+
+    def canonicalize(
+        self,
+        reward_function: types.RewardFunction,
+        /,
+        n_samples_can: Optional[int],
+    ) -> types.RewardFunction:
+        """Canonicalizes a reward function into a divergence-free reward
+        function of the same equivalence class.
+
+        This is done by fitting a potential function (constructed from a neural
+        network) to minimize the L2 norm of the shaped reward function.
+        """
+        rew_fn = utils.multidim_rew_fn(reward_function)
+        n_samples_can = n_samples_can or self.default_samples_can
+        assert isinstance(n_samples_can, int)
+
+        net = nn.Sequential(
+            *[
+                nn.Flatten(),
+                nn.Linear(self.state_dim, max(self.state_dim * 4, 128)),
+                nn.GELU(),
+                Residual(nn.Linear(max(self.state_dim * 4, 128), max(self.state_dim * 4, 128))),
```

Comment:
I think people usually have residual connection over multiple layers (e.g. two linear layers plus the non-linearity in between)? If you just have a linear layer, I wouldn't expect there to be much of a benefit (since it's easy for linear layers to fit an identity), but maybe I'm wrong about that. Anyway, this is just a thought on things to try if the optimization isn't working, I'm aware this specific network is temporary and we'll likely have something more flexible later

Reply:
Yeah makes sense, I think this isn't making a big difference right now (though I changed it to go over both the gelu nonlinearities as well now).